### PR TITLE
feature: send bulk of tx hashes

### PR DESCRIPTION
**Pull request by u2**
pending for benchmark

### CODE REVIEW

File: sync/src/relayer/mod.rs [hunk 1]
```diff
@@ -419,20 +420,61 @@ impl<CS: ChainStore + 'static> Relayer<CS> {
...
+                for tx_hash in tx_hashes {
+                    let peers: Vec<PeerIndex> = nc
+                        .connected_peers()
+                        .into_iter()
+                        .filter(|target_peer| {
+                            known_txs.insert(*target_peer, tx_hash.clone())
+                                && (peer_index != *target_peer)
+                        })
+                        .take(MAX_RELAY_PEERS)
+                        .collect();
```
Comment (quake) [inline sync/src/relayer/mod.rs]: no need to allocate a new vec here, we can use `for_each` after `take` directly

File: sync/src/relayer/transaction_hashes_process.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    }
+
+    pub fn execute(self) -> Result<(), FailureError> {
+        let transaction_hashes: TransactionHashes = (*self.message).try_into()?;
+
+        let mut x_hashes = Vec::with_capacity(transaction_hashes.hashes.len());
+        {
+            let tx_filter = self.relayer.shared().tx_filter();
+            for tx_hash in transaction_hashes.hashes {
```
Comment (quake) [inline sync/src/relayer/transaction_hashes_process.rs]: Can we find a better name for `x_hashes`? naming is hard...
Should use `filter_map` here:

```suggestion
    let x_hashes: Vec<&H256> = transaction_hashes.hashes.iter().filter_map(|tx_hash| {
       ...
    }).collect();
```

File: sync/src/relayer/transaction_hashes_process.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                }
+            }
+        }
+
+        let mut knowned = Vec::with_capacity(x_hashes.len());
+        {
+            let state = self.relayer.shared.lock_chain_state();
+            let tx_pool = state.tx_pool();
+
+            for tx_hash in mem::replace(&mut x_hashes, vec![]) {
```
Comment (quake) [inline sync/src/relayer/transaction_hashes_process.rs]: Suggest to  use `retain` here:
```suggestion
            x_hashes.retain(|tx_hash| {
                ...
            });
```

File: sync/src/relayer/transactions_process.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        TransactionsProcess {
+            message,
+            relayer,
+            nc,
+            peer,
+        }
+    }
+
+    pub fn execute(self) -> Result<(), FailureError> {
+        let relay_txes: RelayTransactions = (*self.message).try_into()?;
```
Comment (quake) [inline sync/src/relayer/transactions_process.rs]: Please use `txs` as `transaction`'s plural form abbreviation:
```suggestion
        let relay_txs: RelayTransactions = (*self.message).try_into()?;
```

File: sync/src/relayer/transactions_process.rs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+        let mut txes = Vec::with_capacity(relay_txes.transactions.len());
+
+        {
+            let tx_filter = self.relayer.shared().tx_filter();
+            for relay_tx in relay_txes.transactions {
```
Comment (quake) [inline sync/src/relayer/transactions_process.rs]: Please use `filter_map` and `collect`

File: sync/src/types.rs [hunk 6]
```diff
@@ -1009,10 +1020,29 @@ impl<CS: ChainStore> SyncSharedState<CS> {
         self.tx_filter.lock().insert(hash);
     }
 
+    pub fn mark_as_known_txes(&self, hashes: Vec<H256>) {
```
Comment (quake) [inline sync/src/types.rs]: Please use `txs` as `transaction`'s plural form abbreviation:

```suggestion
    pub fn mark_as_known_txs(&self, hashes: Vec<H256>) {
```

File: sync/src/relayer/mod.rs (hunk 1 above)
Comment (u2) [reply sync/src/relayer/mod.rs]: updated, thanks.

File: sync/src/relayer/transaction_hashes_process.rs (hunk 2 above)
Comment (u2) [reply sync/src/relayer/transaction_hashes_process.rs]: After changing to `filter_map`, Clippy suggested changing to `filter`.

thanks.

File: sync/src/relayer/transaction_hashes_process.rs (hunk 3 above)
Comment (u2) [reply sync/src/relayer/transaction_hashes_process.rs]: thanks, updated.

File: sync/src/relayer/transactions_process.rs (hunk 4 above)
Comment (u2) [reply sync/src/relayer/transactions_process.rs]: thanks.

File: sync/src/relayer/transactions_process.rs (hunk 5 above)
Comment (u2) [reply sync/src/relayer/transactions_process.rs]: After changing to `fliter_map`, Clippy suggests changing to `filter`.

File: sync/src/types.rs (hunk 6 above)
Comment (u2) [reply sync/src/types.rs]: thanks.

File: shared/src/tx_pool/pool.rs [hunk 7]
```diff
@@ -158,6 +158,14 @@ impl TxPool {
             || self.orphan.contains_key(id)
     }
 
+    pub fn contains_tx(&self, id: &ProposalShortId) -> bool {
+        self.pending.contains_key(id)
+            || self.gap.contains_key(id)
+            || self.proposed.contains_key(id)
+            || self.orphan.contains_key(id)
+            || self.conflict.contains_key(id)
```
Comment (driftluo) [inline shared/src/tx_pool/pool.rs]: I think the `conflict pool` should not be added to the judgment. The tx in the conflict pool may not be updated for a long time, there will be dirty data

File: sync/src/types.rs [hunk 8]
```diff
@@ -1012,10 +1023,29 @@ impl<CS: ChainStore> SyncSharedState<CS> {
...
+            }
+        }
+
+        let mut tx_filter = self.tx_filter.lock();
+
+        for hash in hashes {
+            tx_filter.insert(hash);
+        }
+    }
+
```
Comment (keroro520) [inline sync/src/types.rs]: `mark_as_known_tx` can reuse `mark_as_known_txs`.

```
pub fn mark_as_known_tx(&self, hash: H256) {
    self.mark_as_known_txs(vec![hash])
}
```

File: sync/src/relayer/transaction_hashes_process.rs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        let transaction_hashes: TransactionHashes = (*self.message).try_into()?;
+
+        let mut transit_hashes: Vec<H256> = {
+            let tx_filter = self.relayer.shared().tx_filter();
+            transaction_hashes
+                .hashes
+                .into_iter()
+                .filter(|tx_hash| {
+                    if tx_filter.contains(&tx_hash) {
+                        debug_target!(
```
Comment (keroro520) [inline sync/src/relayer/transaction_hashes_process.rs]: Most of the `transaction_hash` is duplicated. So I would suggest removing this logging.

File: sync/src/relayer/transaction_hashes_process.rs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            .shared()
+            .peers()
+            .state
+            .write()
+            .get_mut(&self.peer)
+        {
+            let mut inflight_transactions = self.relayer.shared().inflight_transactions();
+
+            for tx_hash in transit_hashes {
+                debug_target!(
```
Comment (keroro520) [inline sync/src/relayer/transaction_hashes_process.rs]: Ditto

File: sync/src/relayer/transaction_hashes_process.rs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+                    knowned.push(tx_hash.to_owned());
+                    false
+                } else {
+                    true
+                }
+            })
+        }
+
+        {
+            self.relayer.shared().mark_as_known_txs(knowned);
```
Comment (keroro520) [inline sync/src/relayer/transaction_hashes_process.rs]: I think it is unnecessary to mark known txs here. It should very rarely that "`tx_pool` has a tx but `tx_filter` not". So I would suggest removing this logic or add a log when `!knowned.is_empty()`

File: sync/src/relayer/transactions_process.rs [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+            let tx_filter = self.relayer.shared().tx_filter();
+
+            relay_txs
+                .transactions
+                .into_iter()
+                .filter(|relay_tx| {
+                    let tx_hash = relay_tx.transaction.hash();
+                    if tx_filter.contains(tx_hash) {
+                        debug_target!(
+                            crate::LOG_TARGET_RELAY,
```
Comment (keroro520) [inline sync/src/relayer/transactions_process.rs]: Ditto

File: sync/src/relayer/transaction_hashes_process.rs [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+                    tx_hash,
+                    self.peer,
+                );
+
+                let last_ask_timeout = inflight_transactions.get(&tx_hash).cloned();
+
+                if let Some(next_ask_timeout) =
+                    peer_state.add_ask_for_tx(tx_hash.clone(), last_ask_timeout)
+                {
+                    inflight_transactions.insert(tx_hash.clone(), next_ask_timeout);
```
Comment (keroro520) [inline sync/src/relayer/transaction_hashes_process.rs]: `tx_hash.clone()` -> `tx_hash`

File: sync/src/relayer/transaction_hashes_process.rs (hunk 9 above)
Comment (u2) [reply sync/src/relayer/transaction_hashes_process.rs]: I am not sure, just keep the before logic.

File: sync/src/relayer/transaction_hashes_process.rs (hunk 11 above)
Comment (u2) [reply sync/src/relayer/transaction_hashes_process.rs]: I am not sure, just keep the logic as before.

File: sync/src/types.rs (hunk 8 above)
Comment (u2) [reply sync/src/types.rs]: ok

File: sync/src/relayer/transaction_hashes_process.rs (hunk 9 above)
Comment (u2) [reply sync/src/relayer/transaction_hashes_process.rs]: would remove it later.

File: sync/src/relayer/transaction_hashes_process.rs (hunk 11 above)
Comment (u2) [reply sync/src/relayer/transaction_hashes_process.rs]: agree, the filter is a false negative.

File: sync/src/relayer/transactions_process.rs (hunk 12 above)
Comment (u2) [reply sync/src/relayer/transactions_process.rs]: Updated in the last commit, please review again, thanks.

**Pull request merged**